FirebaseAuth/Sources/Auth/FIRAuth.m

```diff
@@ -564,17 +535,67 @@ - (void)protectedDataInitialization {
           [strongSelf updateCurrentUser:user byForce:NO savingToDisk:NO error:&error];
           self->_lastNotifiedUserToken = user.rawAccessToken;
         } else {
+#if TARGET_OS_IOS || TARGET_OS_TV || TARGET_OS_MACCATALYST
+          if (error.code == FIRAuthErrorCodeKeychainError) {
+            // If there's a keychain error, assume it is due to the keychain being accessed
+            // before the device is unlocked as a result of prewarming, and listen for the
+            // UIApplicationProtectedDataDidBecomeAvailable notification.
+            __weak FIRAuth *innerWeakSelf = strongSelf;
+            strongSelf->_protectedDataDidBecomeAvailableObserver =
+                [[NSNotificationCenter defaultCenter]
+                    addObserverForName:UIApplicationProtectedDataDidBecomeAvailable
+                                object:nil
+                                 queue:nil
+                            usingBlock:^(NSNotification *notification) {
+                              FIRAuth *innerStrongSelf = innerWeakSelf;
+                              [innerStrongSelf protectedDataInitialization];
+                            }];
+          }
+#endif  // TARGET_OS_IOS || TARGET_OS_TV || TARGET_OS_MACCATALYST
```

Comment:
Given the code is duplicated 3 times, I think it would make sense to pull it out into a function. Do we also need to stop listening to the observer inside the block?